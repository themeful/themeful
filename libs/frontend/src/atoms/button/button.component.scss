:host {
  display: block;
}

.button {
  border-radius: 0.25rem;
  border: none;
  padding: 8px 16px;
  min-width: 20px;
  min-height: 20px;
  font-weight: 600;
  display: inline-flex;
  cursor: pointer;
  align-items: center;
  justify-content: center;

  --bg-opacity: 0.5;
  --ggs: 0.5;

  &.active {
    --bg-opacity: 0.75;
  }

  &.primary {
    background-color: rgba(50, 90, 240, var(--bg-opacity));
  }

  &.secondary {
    background-color: rgba(200, 200, 200, var(--bg-opacity));
  }

  &.danger {
    background-color: rgba(255, 100, 100, var(--bg-opacity));
  }

  &:disabled {
    filter: grayscale(100%);
    cursor: default;
  }

  &:focus {
    outline: 0 !important;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.5);
  }

  &:focus:not(:focus-visible) {
    box-shadow: none !important;
  }

  &:hover:not(.active):not(:disabled) {
    filter: brightness(0.75);
  }
}

//   &--icon {
//     padding: 2px;
//     width: 24px;
//     height: 24px;
//   }

//   &--small {
//     padding: 4px 8px;
//     font-weight: 400;
//   }

//   & + .m4-button {
//     margin-left: 4px;
//   }

//   &-group {
//     display: flex;

//     & > * {
//       margin-left: 0 !important;
//       flex-grow: 4;

//       &:not(:first-child) {
//         border-top-left-radius: 0;
//         border-bottom-left-radius: 0;
//       }
//       &:not(:last-child) {
//         border-top-right-radius: 0;
//         border-bottom-right-radius: 0;
//       }
//     }
//   }
// }
