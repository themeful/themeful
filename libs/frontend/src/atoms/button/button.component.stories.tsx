import { h } from '@stencil/core'

export default {
  title: 'Atoms/Buttons',
  args: {
    buttons: [
      { label: 'Primary', args: { type: 'button', kind: 'primary' } },
      { label: 'Secondary', args: { type: 'submit', kind: 'secondary' } },
      { label: 'Danger', args: { type: 'button', kind: 'danger' } },
    ],
  },
}

export const buttons = ({ buttons }): HTMLElement => {
  return (
    <div>
      <h1>Buttons</h1>
      <table style={{ borderSpacing: '1rem' }}>
        <tr>
          <td>
            <h3>default</h3>
          </td>
          <td>
            <h3>active</h3>
          </td>
          <td>
            <h3>disabled</h3>
          </td>
        </tr>
        {buttons.map((button) => {
          return (
            <tr>
              <td>
                <tf-button {...button.args}>{button.label}</tf-button>
              </td>
              <td>
                <tf-button {...button.args} active>
                  {button.label}
                </tf-button>
              </td>
              <td>
                <tf-button {...button.args} disabled>
                  {button.label}
                </tf-button>
              </td>
            </tr>
          )
        })}
      </table>
      <h1>With Icon</h1>
      <table>
        <tr>
          <td>default</td>
          <td>
            <tf-button kind="secondary">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
        <tr>
          <td>small</td>
          <td>
            <tf-button kind="secondary" size="small">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="small">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="small">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
        <tr>
          <td>icon</td>
          <td>
            <tf-button kind="secondary" size="icon">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="icon">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="icon">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
      </table>
    </div>
  )
}
