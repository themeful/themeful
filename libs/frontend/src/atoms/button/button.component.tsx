import { Component, Event, h, EventEmitter, Prop } from '@stencil/core'
import { ButtonKind } from './interfaces'

@Component({
  tag: 'tf-button',
  styleUrl: 'button.component.scss',
})
export class TfButtonComponent {
  /** Button type*/
  @Prop() type = 'button'

  /** Button kind */
  @Prop() kind: ButtonKind = 'primary'

  /** Active state */
  @Prop() active: false

  /** Disabled state */
  @Prop() disabled: false

  /** Event emitted when the item is clicked */
  @Event({ bubbles: true, composed: true }) clicked: EventEmitter

  render() {
    return (
      <button
        type={this.type}
        class={`button ${this.kind}${this.active ? ' active' : ''}`}
        disabled={this.disabled}
        onClick={() => {
          this.clicked.emit()
        }}
      >
        <slot />
      </button>
    )
  }
}
