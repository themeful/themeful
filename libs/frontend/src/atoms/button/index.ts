export * from './button.component'
export * from './interfaces'
