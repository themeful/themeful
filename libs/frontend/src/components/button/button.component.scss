:host {
  display: inline-block;
}

.button {
  color: var(--tf-font-color);
  border-radius: 0.25rem;
  border: none;
  padding: 0.5rem 1rem;
  min-width: 20px;
  min-height: 20px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.25rem;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1rem;

  --bg-opacity: 0.5;
  --ggs: 0.5;

  &.small {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.25rem 0.5rem;
  }

  &.icon {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.125rem;
  }

  &.active {
    --bg-opacity: 0.75;
  }

  &.primary {
    background-color: rgba(50, 90, 240, var(--bg-opacity));
  }

  &.secondary {
    background-color: rgba(200, 200, 200, var(--bg-opacity));
  }

  &.danger {
    background-color: rgba(255, 100, 100, var(--bg-opacity));
  }

  &:disabled {
    filter: grayscale(100%);
    cursor: default;
  }

  &:focus {
    outline: 0 !important;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.5);
  }

  &:focus:not(:focus-visible) {
    box-shadow: none !important;
  }

  &:hover:not(.active):not(:disabled) {
    filter: brightness(0.75);
  }
}

//   & + .m4-button {
//     margin-left: 4px;
//   }

//   &-group {
//     display: flex;

//     & > * {
//       margin-left: 0 !important;
//       flex-grow: 4;

//       &:not(:first-child) {
//         border-top-left-radius: 0;
//         border-bottom-left-radius: 0;
//       }
//       &:not(:last-child) {
//         border-top-right-radius: 0;
//         border-bottom-right-radius: 0;
//       }
//     }
//   }
// }
