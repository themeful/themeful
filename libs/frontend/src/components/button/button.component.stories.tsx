import { h } from '@stencil/core'
import Fragment from 'stencil-fragment'

export default {
  title: 'Components/Buttons',
  args: {
    buttons: [
      { label: 'Primary', args: { type: 'button', kind: 'primary' } },
      { label: 'Secondary', args: { type: 'submit', kind: 'secondary' } },
      { label: 'Danger', args: { type: 'button', kind: 'danger' } },
      { label: 'Selectable', args: { type: 'button', kind: 'selectable' } },
    ],
  },
}

const renderButtons = (buttons) => (
  <Fragment>
    <table style={{ borderSpacing: '1rem' }}>
      <tr>
        <td>
          <h3>default</h3>
        </td>
        <td>
          <h3>active</h3>
        </td>
        <td>
          <h3>disabled</h3>
        </td>
      </tr>
      {buttons.map((button) => {
        return (
          <tr>
            <td>
              <tf-button {...button.args}>{button.label}</tf-button>
            </td>
            <td>
              <tf-button {...button.args} active>
                {button.label}
              </tf-button>
            </td>
            <td>
              <tf-button {...button.args} disabled>
                {button.label}
              </tf-button>
            </td>
          </tr>
        )
      })}
    </table>
    <h1>With Icon</h1>
    <table>
      <tr>
        <td>default</td>
        <td>
          <tf-button>
            <tf-icon icon="sync" />
            Some Text
          </tf-button>
        </td>
        <td>
          <tf-button>
            Some Text
            <tf-icon icon="sync" />
          </tf-button>
        </td>
        <td>
          <tf-button>
            <tf-icon icon="sync" />
          </tf-button>
        </td>
      </tr>
      <tr>
        <td>small</td>
        <td>
          <tf-button size="small">
            <tf-icon icon="sync" />
            Some Text
          </tf-button>
        </td>
        <td>
          <tf-button size="small">
            Some Text
            <tf-icon icon="sync" />
          </tf-button>
        </td>
        <td>
          <tf-button size="small">
            <tf-icon icon="sync" />
          </tf-button>
        </td>
      </tr>
      <tr>
        <td>icon</td>
        <td>
          <tf-button size="icon">
            <tf-icon icon="sync" />
            Some Text
          </tf-button>
        </td>
        <td>
          <tf-button size="icon">
            Some Text
            <tf-icon icon="sync" />
          </tf-button>
        </td>
        <td>
          <tf-button size="icon">
            <tf-icon icon="sync" />
          </tf-button>
        </td>
      </tr>
    </table>
  </Fragment>
)

export const buttons = ({ buttons }): HTMLElement => {
  return (
    <div>
      <div class="header">
        <h1>Buttons</h1>
      </div>
      <div class="content content--grid">
        <div class="tf-light">{renderButtons(buttons)}</div>
        <div>{renderButtons(buttons)}</div>
      </div>
    </div>
  )
}
