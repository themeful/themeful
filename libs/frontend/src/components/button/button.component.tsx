import { Component, h, Prop } from '@stencil/core'
import { ButtonKind } from './interfaces'

@Component({
  tag: 'tf-button',
  styleUrl: 'button.component.scss',
})
export class ButtonComponent {
  /** Button type*/
  @Prop() type = 'button'

  /** Button kind */
  @Prop() kind: ButtonKind = 'secondary'

  /** Button size */
  @Prop() size: 'small' | 'large' | 'icon'

  /** Active state */
  @Prop() active: false

  /** Disabled state */
  @Prop() disabled: false

  public render(): HTMLTfButtonElement {
    return (
      <button
        type={this.type}
        class={`button ${this.kind} ${this.size ?? ''}${this.active ? ' active' : ''}`}
        disabled={this.disabled}
      >
        <slot />
      </button>
    )
  }
}
