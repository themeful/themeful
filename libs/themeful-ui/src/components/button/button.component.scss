:host {
  display: inline-block;
}

.button {
  color: var(--tf-font-color);
  border-radius: 0.25rem;
  border: none;
  padding: 0.5rem 1rem;
  min-width: 20px;
  width: 100%;
  min-height: 20px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.25rem;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1rem;

  --bg-opacity: 0.5;
  --ggs: 0.5;

  &.small {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.25rem 0.5rem;
  }

  &.icon {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.125rem;
  }

  &.active {
    --bg-opacity: 0.75;
  }

  &.selectable:not(.active),
  &.secondary {
    background-color: rgba(200, 200, 200, var(--bg-opacity));
  }

  :host-context(.link-active) &.selectable,
  &.selectable.active,
  &.primary {
    background-color: rgba(50, 90, 240, var(--bg-opacity));
  }

  &.danger {
    background-color: rgba(255, 100, 100, var(--bg-opacity));
  }

  &:disabled {
    filter: grayscale(100%);
    cursor: default;
  }

  &:focus {
    outline: 0 !important;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.5);
  }

  &:focus:not(:focus-visible) {
    box-shadow: none !important;
  }

  &:hover:not(.active):not(:disabled) {
    filter: brightness(0.75);
  }
}

:host-context(.button-group) {
  .button {
    border-radius: 0;
  }
}

:host(.button-group--first) {
  .button {
    border-top-left-radius: 0.5rem;
    border-bottom-left-radius: 0.5rem;
  }
}

:host(.button-group--last) {
  .button {
    border-top-right-radius: 0.5rem;
    border-bottom-right-radius: 0.5rem;
  }
}
