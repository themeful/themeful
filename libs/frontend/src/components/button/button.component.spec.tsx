import { newSpecPage } from '@stencil/core/testing'
import { ButtonComponent } from './button.component'

describe('tf-button', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [ButtonComponent],
      html: '<tf-button />',
    })
    expect(root).toEqualHtml(`
      <tf-button>
        <mock:shadow-root>
          <div>
            Hello, World! I'm
          </div>
        </mock:shadow-root>
      </tf-button>
    `)
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [ButtonComponent],
      html: `<tf-button first="Stencil" last="'Don't call me a framework' JS" />`,
    })
    expect(root).toEqualHtml(`
      <tf-button first="Stencil" last="'Don't call me a framework' JS">
        <mock:shadow-root>
          <div>
            Hello, World! I'm Stencil 'Don't call me a framework' JS
          </div>
        </mock:shadow-root>
      </tf-button>
    `)
  })
})
