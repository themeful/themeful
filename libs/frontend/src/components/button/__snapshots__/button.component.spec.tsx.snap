// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonComponent renders 1`] = `
<tf-button>
  <mock:shadow-root>
    <button class="button secondary" type="button">
      <slot></slot>
    </button>
  </mock:shadow-root>
</tf-button>
`;

exports[`ButtonComponent renders with values 1`] = `
<tf-button>
  <mock:shadow-root>
    <button class="button primary" type="submit">
      <slot></slot>
    </button>
  </mock:shadow-root>
  Hello World
</tf-button>
`;
